### PR TITLE
Make sync work across VV upgrades

### PR DESCRIPTION
**Pull request by snej**
The transition from tree-based revids to version vectors is tricky for replication, and the prior approach (adding a fake 'legacy' version when creating the vector) didn't turn out to work.

Instead, now:
1. when a doc is upgraded in memory to a VectorRecord it keeps its tree-based/legacy revids. 
2. When the doc is saved, it creates a version vector for the local Revision but also remembers the prior legacy revid. 
3. Remote Revisions keep their tree-based IDs until they're replaced by newer VV revisions (pulled or pushed).

As a result, revision histories now consist of a VV and/or legacy revids. The version vector, if present, always comes first. The legacy IDs at the end are what allow the recipient of a revision (an active or passive puller) to match up an incoming version vector with a local legacy revid.

(It's also possible for a rev history sent by the replicator to consist _only_ of legacy IDs, even after the upgrade. This happens if a doc was updated locally before the app upgrade but not pushed until after. It can also happen on a pull for the same reason.)

There's an open question of how long we need to keep the legacy revid of the document. It only needs to be used once per remote, but in a P2P scenario you never know if you might have to sync with some new peer that hasn't upgraded that doc yet. In any case the revid isn't very big (~24 bytes) and it only occurs in pre-existing docs, so it may not be worth removing.

### CODE REVIEW

**Comment by jianminzhao** [comment]
yes, that test is flaky. It seems to happen when the replicator goes to Idle before all revisions get synced.

**Comment by jianminzhao** [comment]
"when a doc is upgraded in memory to a VectorRecord" -- I noticed that method "upgradeVersioning" is deleted. When actually a doc is upgraded in memory?

**Comment by pasin** [comment]
Reviewing the code changes from looking at the diff here is hard for me to understand all of upgrading and handling both types of revisions. I'm still working on it. As of now, I have a few questions in general: 

> when a doc is upgraded in memory to a VectorRecord it keeps its tree-based/legacy revids.

Do I understand here that if the doc didn't get upgraded to vv, the vv record will be generated when read (but will not be saved)?

> (It's also possible for a rev history sent by the replicator to consist only of legacy IDs, even after the upgrade. This happens if a doc was updated locally before the app upgrade but not pushed until after. It can also happen on a pull for the same reason.)

Do I understand correctly that when pulling a new revision that only contains the rev tree, the doc will be upgraded to the version vector (with the old rev tree kept) at that point including the new doc?

**Comment by snej** [comment]
> "when a doc is upgraded in memory to a VectorRecord" -- I noticed that method "upgradeVersioning" is deleted. When actually a doc is upgraded in memory?

Any time it's opened by the current version of LiteCore; the DocumentFactory is now VectorDocumentFactory.

File: LiteCore/Database/VectorDocument.cc [hunk 1]
```diff
@@ -196,25 +197,51 @@ namespace litecore {
...
-                            auto parent = _doc.remoteRevision(parID);
-                            if ( parent.value().revID != rev->revID ) { history += ", " + parent->revID.str(); }
+                        // Start with the doc's last legacy rev:
+                        if ( revid lastRevID = _doc.lastLegacyRevID() ) vvl.legacy.emplace_back(lastRevID);
+                    }
+
+                    // Search the remotes for earlier legacy revs that aren't conflicts:
+                    unsigned curGen = UINT_MAX;
+                    if ( !vvl.legacy.empty() ) curGen = revid(vvl.legacy.back()).generation();
+                    revid ancestorRevID;
```
[the remark this answers is not in the document]
Comment (snej) [reply LiteCore/Database/VectorDocument.cc]: Removed it!

**Review by pasin — APPROVED** [review]

**Comment by snej** [comment]
Please review the last commit, which is a change I had to make to get a unit test pass. I think the code in the test was wrong.

**Review by jianminzhao — APPROVED** [review]

**Pull request merged**